### PR TITLE
Policy Enforcement Fixes

### PR DESCRIPTION
**Pull request by ianvernon**
This commit addresses various issues with respect to Policy Enforcement, specifically those mentioned in the below tickets. 

Policy enforcement mode is now configurable via the CLI in the following manner:
`cilium config PolicyEnforcement={always, default, enabled}`. 

Before, it was: `cilium config Policy={Enabled, Disabled}`

This meant that it was  impossible to switch Policy Enforcement mode back to the 'default' mode (for more information w.r.t. the specifics for policy enforcement see issue #788 ). 

Extensive tests have been added to ensure that the configuration updates for policy enforcement are reflected in the endpoints managed by Cilium. 

Fixes: #1298 
Fixes: #1291 
Fixes: #1082

### CODE REVIEW

File: cilium/cmd/helpers.go [hunk 1]
```diff
@@ -20,6 +20,7 @@ import (
 
 	"github.com/cilium/cilium/pkg/endpoint"
 
+	"github.com/cilium/cilium/pkg/policy"
```
Comment (aanm) [inline cilium/cmd/helpers.go]: should be one line above

File: cilium/cmd/endpoint_list.go [hunk 2]
```diff
@@ -44,7 +44,7 @@ func init() {
 
 func listEndpoint(w *tabwriter.Writer, ep *models.Endpoint, id string, label string) {
 	var isPolicyEnabled string
-	if ep.PolicyEnabled {
+	if *ep.PolicyEnabled {
```
Comment (aanm) [inline cilium/cmd/endpoint_list.go]: is this really needed?

File: cilium/cmd/config.go [hunk 3]
```diff
@@ -22,9 +22,10 @@ import (
 	"github.com/cilium/cilium/api/v1/models"
 	"github.com/cilium/cilium/common"
 	"github.com/cilium/cilium/daemon/options"
-	"github.com/cilium/cilium/pkg/option"
 
+	"github.com/cilium/cilium/pkg/option"
```
Comment (aanm) [inline cilium/cmd/config.go]: imports out of order

File: cilium/cmd/config.go [hunk 4]
```diff
@@ -77,12 +78,26 @@ func configDaemon(cmd *cobra.Command, opts []string) {
 		dumpConfig(resp.Configuration.Mutable)
 		fmt.Printf("%-24s %s\n", "k8s-configuration", resp.K8sConfiguration)
 		fmt.Printf("%-24s %s\n", "k8s-endpoint", resp.K8sEndpoint)
+		fmt.Printf("%-24s %s\n", "PolicyEnforcement", resp.PolicyEnforcement)
```
Comment (aanm) [inline cilium/cmd/config.go:81]: We already have a `Policy` flag. What will be the meaning of this one? I think we should rename it to `PolicyEnforcementMode` or `PolicyMode`
```
$ cilium config
Conntrack                Enabled
ConntrackAccounting      Enabled
ConntrackLocal           Disabled
Debug                    Enabled
DropNotification         Enabled
Policy                   Disabled
PolicyTracing            Disabled
k8s-configuration        /var/lib/cilium/cilium.kubeconfig
k8s-endpoint             http://
```

File: daemon/daemon.go [hunk 5]
```diff
@@ -244,11 +244,9 @@ func (d *Daemon) AlwaysAllowLocalhost() bool {
 	return d.conf.alwaysAllowLocalhost
 }
 
-func (d *Daemon) PolicyEnabled() bool {
-	return d.conf.Opts.IsEnabled(endpoint.OptionPolicy)
-}
-
 func (d *Daemon) PolicyEnforcement() string {
+	d.conf.EnablePolicyMU.RLock()
+	defer d.conf.EnablePolicyMU.RUnlock()
```
Comment (aanm) [inline daemon/daemon.go]: `defer d.conf.EnablePolicyMU.RUnlock()` creates overhead for a small function like this, you can write it as 
```
func (d *Daemon) PolicyEnforcement() (pe string) {
     d.conf.EnablePolicyMU.RLock()
     pe = d.conf.EnablePolicy
     d.conf.EnablePolicyMU.RUnlock()
     return
}
```

**Review by aanm — COMMENTED** [review]
Note to self: reviewed until `daemon: PolicyEnforcement fixes`

File: api/v1/openapi.yaml [hunk 6]
```diff
@@ -805,6 +808,20 @@ definitions:
         type: array
         items:
           type: string
+  L3Policy:
```
Comment (tgraf) [inline api/v1/openapi.yaml]: Can we rename this to CIDRPolicy? As I understand this, this will only contain the CIDR level policy rules

File: api/v1/openapi.yaml [hunk 7]
```diff
@@ -805,6 +808,20 @@ definitions:
...
+    description: L3 endpoint policy
+    type: object
+    properties:
+      ingress:
+        description: list of L3 Ingress rules
```
Comment (tgraf) [inline api/v1/openapi.yaml]: Description should also make it clear that these are CIDR rules

File: cilium/cmd/endpoint_list.go (hunk 2 above)
Comment (tgraf) [reply cilium/cmd/endpoint_list.go]: This is not safe. If it's a pointer then swagger will not guarantee that it is always set.

You can do:
```
import (
             "github.com/go-openapi/swag"
)

if swag.BoolValue(ep.PolicyEnabled) {
```

File: cilium/cmd/config.go (hunk 4 above)
Comment (tgraf) [reply cilium/cmd/config.go:81]: Can we just remove the old Policy know given we put a more powerful knob in place?

File: daemon/daemon.go [hunk 8]
```diff
@@ -1086,23 +1084,56 @@ func (h *patchConfig) Handle(params PatchConfigParams) middleware.Responder {
...
+	config.EnablePolicyMU.RLock()
+	oldEnforcementValue := config.EnablePolicy
+	config.EnablePolicyMU.RUnlock()
+	enforcement := params.Configuration.PolicyEnforcement
+
+	// Only update if value provided for PolicyEnforcement.
+	if enforcement != "" {
+		switch enforcement {
+		case endpoint.NeverEnforce, endpoint.DefaultEnforcement, endpoint.AlwaysEnforce:
+			// If the policy enforcement configuration has indeed changed, we have
```
Comment (tgraf) [inline daemon/daemon.go]: You have a race condition here. Because you unlock `EnablePolicyMU` and keep the variable it may change by the time you compare it.

I suggest you lock here...

File: daemon/daemon.go [hunk 9]
```diff
@@ -1086,23 +1084,56 @@ func (h *patchConfig) Handle(params PatchConfigParams) middleware.Responder {
...
+				// Only allow configuration updates for PolicyEnforcement to occur again once configuration has been applied
+				config.EnablePolicyMU.Lock()
+				defer config.EnablePolicyMU.Unlock()
+				config.EnablePolicy = enforcement
+
+				d.GetPolicyRepository().Mutex.RLock()
+				_, policyEnforcementChanged = d.EnablePolicyEnforcement()
+				d.GetPolicyRepository().Mutex.RUnlock()
+				d.TriggerPolicyUpdates([]policy.NumericIdentity{})
+			}
```
Comment (tgraf) [inline daemon/daemon.go]: ... and unlock here

File: daemon/daemon.go [hunk 10]
```diff
@@ -1086,23 +1084,56 @@ func (h *patchConfig) Handle(params PatchConfigParams) middleware.Responder {
...
+			if enforcement != oldEnforcementValue {
+				// Only allow configuration updates for PolicyEnforcement to occur again once configuration has been applied
+				config.EnablePolicyMU.Lock()
+				defer config.EnablePolicyMU.Unlock()
+				config.EnablePolicy = enforcement
+
+				d.GetPolicyRepository().Mutex.RLock()
+				_, policyEnforcementChanged = d.EnablePolicyEnforcement()
+				d.GetPolicyRepository().Mutex.RUnlock()
+				d.TriggerPolicyUpdates([]policy.NumericIdentity{})
```
Comment (tgraf) [inline daemon/daemon.go]: You can move the `TriggerPolicyUpdates()` to the `if changes > 0` branch

**Review by tgraf — CHANGES_REQUESTED** [review]

File: cilium/cmd/config.go (hunk 4 above)
Comment (ianvernon) [reply cilium/cmd/config.go:81]: I'll see if I can remove the old Policy boolean option entirely from the daemon. I only think it is used in the daemon in `daemon/daemon.go:writeNetdevHeader()`. I can create a function that writes the `#define` for PolicyEnforcement for the daemon.

File: daemon/daemon.go (hunk 10 above)
Comment (ianvernon) [reply daemon/daemon.go]: That means that upon any configuration update to the daemon, we might have to regenerate each endpoint as opposed to only if PolicyEnforcement is changed - is this desirable? I figured we'd keep it here so we don't have to regenerate if needed, as regeneration is costly.

File: daemon/daemon.go (hunk 9 above)
Comment (ianvernon) [reply daemon/daemon.go]: I fixed the race condition, but not deferring the unlocking of `EnablePolicyMU` introduced very weird behavior on my end, including putting endpoints into 'not-ready' state due to compilation errors when trying to regenerate the endpoints. I can reproduce this fairly easily by changing the locking logic. I checked the generated header files / C code when I saw that the endpoints couldn't regenerate and I couldn't figure out why the compiler was complaining.

File: cilium/cmd/config.go (hunk 4 above)
Comment (ianvernon) [reply cilium/cmd/config.go:81]: Done.

File: api/v1/openapi.yaml (hunk 6 above)
Comment (ianvernon) [reply api/v1/openapi.yaml]: Done.

File: api/v1/openapi.yaml (hunk 7 above)
Comment (ianvernon) [reply api/v1/openapi.yaml]: Done.

File: daemon/daemon.go (hunk 8 above)
Comment (ianvernon) [reply daemon/daemon.go]: Fixed - see comment on unlocking of the mutex as well.

File: cilium/cmd/endpoint_list.go (hunk 2 above)
Comment (ianvernon) [reply cilium/cmd/endpoint_list.go]: Fixed.

File: cilium/cmd/helpers.go (hunk 1 above)
Comment (ianvernon) [reply cilium/cmd/helpers.go]: Fixed.

File: cilium/cmd/config.go (hunk 3 above)
Comment (ianvernon) [reply cilium/cmd/config.go]: Fixed.

File: daemon/policy.go [hunk 11]
```diff
@@ -77,46 +76,65 @@ func (d *Daemon) TriggerPolicyUpdates(added []policy.NumericIdentity) *sync.Wait
...
+	//   should be applied to the specified endpoint.
+	// * if we are not running K8s and are running in 'default' mode, we do not
+	//   enable policy enforcement on a per-endpoint basis (i.e., outside of the
+	//   scope of this function).
 	return false
 }
 
 // EnablePolicyEnforcement returns whether policy enforcement needs to be
-// enabled for the daemon.
+// enabled at the daemon-level.
```
Comment (aanm) [inline daemon/policy.go:114]: Can you update this comment? it returns 2 booleans, not one.

File: daemon/policy.go [hunk 12]
```diff
@@ -77,46 +76,65 @@ func (d *Daemon) TriggerPolicyUpdates(added []policy.NumericIdentity) *sync.Wait
...
-			d.conf.Opts.Set(endpoint.OptionPolicy, false)
-			return false
+			//d.conf.Opts.Set(endpoint.OptionPolicy, false)
+			config[endpoint.OptionPolicy] = "disabled"
+			changes := d.conf.Opts.Apply(config, changedOption, d)
+			return false, changes > 0
 		}
 	}
-	return false
+	config[endpoint.OptionPolicy] = "disabled"
```
Comment (aanm) [inline daemon/policy.go]: A comment to known when would this line be reached would help.

File: daemon/policy.go [hunk 13]
```diff
@@ -77,46 +76,65 @@ func (d *Daemon) TriggerPolicyUpdates(added []policy.NumericIdentity) *sync.Wait
 		// FIXME: Invalidate only cache that is affected
 		d.invalidateCache()
 	}
-
-	d.GetPolicyRepository().Mutex.RLock()
-	d.EnablePolicyEnforcement()
```
Comment (aanm) [inline daemon/policy.go:82]: Why was this removed?

File: pkg/endpoint/owner.go [hunk 14]
```diff
@@ -27,15 +27,12 @@ type Owner interface {
 	// Must return true if dry mode is enabled
 	DryModeEnabled() bool
 
-	// PolicyEnabled returns whether policy enforcement is enabled
-	PolicyEnabled() bool
-
 	// EnablePolicyEnforcement returns whether owner should enable policy enforcement.
```
Comment (aanm) [inline pkg/endpoint/owner.go:30]: Update comment accordingly with the number of boolean returned

File: tests/k8s/tests/00-gsg-test.sh [hunk 15]
```diff
@@ -25,10 +25,10 @@ K8SDIR="${dir}/../../../examples/kubernetes"
 GSGDIR="${dir}/deployments/gsg"
 
 function cleanup {
-	kubectl delete -f "${MINIKUBE}/l3_l4_l7_policy.yaml" 2> /dev/null || true
-	kubectl delete -f "${MINIKUBE}/l3_l4_policy_deprecated.yaml" 2> /dev/null || true
-	kubectl delete -f "${MINIKUBE}/l3_l4_policy.yaml" 2> /dev/null || true
-	kubectl delete -f "${GSGDIR}/demo.yaml" 2> /dev/null || true
+  kubectl delete -f "${MINIKUBE}/l3_l4_l7_policy.yaml" 2> /dev/null || true
```
Comment (aanm) [inline tests/k8s/tests/00-gsg-test.sh:28]: This could be made on a different PR, it's not easy to review 30 commits on a single PR

File: tests/helpers.bash [hunk 16]
```diff
@@ -235,6 +235,17 @@ function wait_for_running_pod {
     set -x
 }
 
+function wait_for_no_pods {
+  set +x
+  namespace=${2:-default}
```
Comment (aanm) [inline tests/helpers.bash]: no_pods=$1

File: tests/helpers.bash [hunk 17]
```diff
@@ -235,6 +235,17 @@ function wait_for_running_pod {
...
+  echo "Waiting for no pods to be Running in namespace ${namespace}"
+  kubectl get pods -n ${namespace} -o wide
+  while [[ "$(kubectl get pods -n ${namespace} -o wide 2>&1 | grep -c 'No resources found')" -ne "1" ]] ; do
```
Comment (aanm) [inline tests/helpers.bash]: `-ne "${no_pods}"`

File: tests/14-policy-enforcement-docker.sh [hunk 18]
```diff
@@ -87,7 +87,7 @@ function import_sample_policy {
 [{
      "endpointSelector": {"matchLabels":{"k8s:id":"app3"}},
      "egress": [{
-         "toCIDR": [ { "ip": "9.9.9.9/32" } ]
+         "toCIDR": [ "9.9.9.9/32" ]
```
Comment (aanm) [inline tests/14-policy-enforcement-docker.sh]: This can be squashed with ab2dd4ead4fb30ab4baeb5acd32dd13b19d80c8e (2 commits before)

File: daemon/policy.go (hunk 13 above)
Comment (ianvernon) [reply daemon/policy.go:82]: It's out of place here given how this patch changes how we handle configuration updates. his is handled as part of the handler for configuration changes, which calls `d.EnablePolicyEnforcement` before it calls `TriggerPolicyUpdates`. Since `TriggerPolicyUpdates` is triggered when the daemon's configuration is changed for `PolicyEnforcement`. `EnablePolicyEnforcement()` doesn't actually change any configuration in the daemon any more in this patch, so running it has no effect here anyway.

File: daemon/policy.go (hunk 11 above)
Comment (ianvernon) [reply daemon/policy.go:114]: This is in one of the earlier commits in the branch - I have reverted this functionality in a later commit. I'm going to re-order the commits in the branch, and will re-push once that's done

File: daemon/daemon.go (hunk 9 above)
Comment (ianvernon) [reply daemon/daemon.go]: Resolved this offline with @tgraf.

File: tests/helpers.bash (hunk 17 above)
Comment (ianvernon) [reply tests/helpers.bash]: line 240 should be `namespace=${1:-default}`, but I don't see the need for ${no_pods}. On line 243, I am checking that the number of lines containing "No resources found" is 1, which is what is shown when there are no pods running in a namespace.

File: tests/14-policy-enforcement-docker.sh (hunk 18 above)
Comment (ianvernon) [reply tests/14-policy-enforcement-docker.sh]: I've squashed the test changes into one commit.

File: daemon/daemon.go [hunk 19]
```diff
@@ -322,10 +327,24 @@ func (d *Daemon) writeNetdevHeader(dir string) error {
...
+	fw.WriteString(d.fmtPolicyEnforcement())
 
 	return fw.Flush()
 }
 
+// returns #define for PolicyEnforcement based on the configuration of the daemon.
+// Must be called with d.conf.EnablePolicyMU locked.
+func (d *Daemon) fmtPolicyEnforcement() string {
+	d.GetPolicyRepository().Mutex.RLock()
+	enforcement := d.EnablePolicyEnforcement()
```
Comment (ianvernon) [inline daemon/daemon.go:339]: Note: this always returns false in the case of when Cilium is running in tandem with K8s in 'default' mode, meaning that we always allow egress traffic to the host. Discussed offline with @tgraf , and we both agree that this isn't trivial to deal with at this time.

File: daemon/policy.go (hunk 12 above)
Comment (ianvernon) [reply daemon/policy.go]: Comment has been added.

File: pkg/endpoint/owner.go (hunk 14 above)
Comment (ianvernon) [reply pkg/endpoint/owner.go:30]: This is no longer relevant with the latest commits for this branch.

File: api/v1/openapi.yaml [hunk 20]
```diff
@@ -702,6 +703,8 @@ definitions:
       policy-revision:
         description: The policy revision this endpoint is running on
         type: integer
+      cidr-policy:
```
Comment (tgraf) [inline api/v1/openapi.yaml]: Can we move this into `EndpointPolicy` instead?

File: daemon/policy.go [hunk 21]
```diff
@@ -77,45 +76,59 @@ func (d *Daemon) TriggerPolicyUpdates(added []policy.NumericIdentity) *sync.Wait
...
-	if d.EnablePolicyEnforcement() {
+// Must be called with e.Consumable.Mutex and d.GetPolicyRepository().Mutex held.
+func (d *Daemon) EnableEndpointPolicyEnforcement(e *endpoint.Endpoint) bool {
+	// First check if policy enforcement should be enabled at the daemon level.
+	// If policy enforcement is enabled for the daemon, then it has to be
+	// enabled for the endpoint.
+
+	config.EnablePolicyMU.RLock()
+	daemonPolicyEnable := d.EnablePolicyEnforcement()
+	defer config.EnablePolicyMU.RUnlock()
```
Comment (tgraf) [inline daemon/policy.go]: Can you remove the defer one line up?

**Review by tgraf — CHANGES_REQUESTED** [review]
Looks good otherwise

**Review by tgraf — APPROVED** [review]
